In [1]:
#import libraries 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sqlite3

In [2]:
#connect to the MIMIC database using sqlite3 library.
con = sqlite3.connect('/Users/aneruthmohanasundaram/Documents/GitHub/Data-Analytics/Practical Session 2/Datasets/mimicdata.sqlite')

In [3]:
#list all the table names present in the database.
cursor = con.cursor()
columns = [col[0] for col in cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")]
print(columns)

['CHARTEVENTS', 'DIAGNOSES_ICD', 'D_ICD_DIAGNOSES', 'D_ITEMS', 'D_LABITEMS', 'ICUSTAYS', 'INPUTEVENTS_MV', 'LABEVENTS', 'OUTPUTEVENTS', 'PRESCRIPTIONS', 'PROCEDUREEVENTS_MV', 'TRANSFERS', 'ADMISSIONS', 'MICROBIOLOGYEVENTS', 'PATIENTS']


In [4]:
#choose two tables of your choice and find out for each of them, their corresponding column names and the number of records present.

# For Table 1
cursor.execute("SELECT * FROM DIAGNOSES_ICD")
table1 = cursor.fetchall()
print(f'Number of records present in table1 {len(table1)}')

# For Table 2
cursor.execute("SELECT * FROM ICUSTAYS")
table2 = cursor.fetchall()
print(f'Number of records present in table2 {len(table2)}')

Number of records present in table1 183
Number of records present in table2 11


In [5]:
#read these two tables into a dataframe.
query = """SELECT *
        FROM DIAGNOSES_ICD,ICUSTAYS """
df = pd.read_sql_query(query,con)

In [6]:
# To print dataframe head
df.head()

ROW_ID  SUBJECT_ID  HADM_ID  SEQ_NUM ICD9_CODE  ROW_ID  SUBJECT_ID  \
0  379265       40004   121157        1     34591   41339       40004   
1  379265       40004   121157        1     34591   41347       40036   
2  379265       40004   121157        1     34591   41363       40080   
3  379265       40004   121157        1     34591   41365       40084   
4  379265       40004   121157        1     34591   41373       40116   

   HADM_ID  ICUSTAY_ID    DBSOURCE FIRST_CAREUNIT LAST_CAREUNIT  FIRST_WARDID  \
0   121157      256307  metavision           MICU          MICU            50   
1   198489      285485  metavision          MSICU         MSICU            52   
2   162107      252522  metavision           MICU          MICU            23   
3   195762      264630  metavision          TSICU         TSICU            14   
4   157106      232646  metavision           MICU          MICU            23   

   LAST_WARDID               INTIME              OUTTIME     LOS  
0           50  2171-03-01 17:51:30  2171-03-05 18:28:23  4.0256  
1           52  2141-08-01 23:48:48  2141-08-02 18:41:50  0.7868  
2           23  2106-05-31 16:43:46  2106-06-05 13:18:50  4.8577  
3           14  2173-01-31 22:11:54  2173-02-05 04:13:20  4.2510  
4           23  2150-02-19 00:14:33  2150-02-25 17:03:04  6.7004

In [18]:
#display only the microbiology_events information for which a blood culture was performed.
df_micro = pd.read_sql_query(""" SELECT * FROM 'MICROBIOLOGYEVENTS'""",con) #returns dataframe
bloodcult = df_micro[df_micro['SPEC_TYPE_DESC'] == 'BLOOD CULTURE']
bloodcult

ROW_ID  SUBJECT_ID  HADM_ID            CHARTDATE            CHARTTIME  \
13  268050       40161   180396  2181-10-24 00:00:00  2181-10-24 18:10:00   
16  268847       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
17  268848       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
18  268849       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
19  268850       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
20  268851       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
21  268852       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
22  268853       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
23  268854       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
24  268855       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
76  268557       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
77  268558       40161   180396  2181-11-23 00:00:00  2181-11-23 23:47:00   
90  315552       40004   121157  2171-03-04 00:00:00  2171-03-04 21:12:00   

    SPEC_ITEMID SPEC_TYPE_DESC  ORG_ITEMID  \
13        70012  BLOOD CULTURE       80238   
16        70012  BLOOD CULTURE       80005   
17        70012  BLOOD CULTURE       80005   
18        70012  BLOOD CULTURE       80005   
19        70012  BLOOD CULTURE       80005   
20        70012  BLOOD CULTURE       80005   
21        70012  BLOOD CULTURE       80005   
22        70012  BLOOD CULTURE       80005   
23        70012  BLOOD CULTURE       80005   
24        70012  BLOOD CULTURE       80005   
76        70012  BLOOD CULTURE       80005   
77        70012  BLOOD CULTURE       80005   
90        70012  BLOOD CULTURE       80155   

                                             ORG_NAME  ISOLATE_NUM  AB_ITEMID  \
13  CLOSTRIDIUM SPECIES NOT C. PERFRINGENS OR C. S...            1        NaN   
16                                 KLEBSIELLA OXYTOCA            1    90012.0   
17                                 KLEBSIELLA OXYTOCA            1    90013.0   
18                                 KLEBSIELLA OXYTOCA            1    90017.0   
19                                 KLEBSIELLA OXYTOCA            1    90018.0   
20                                 KLEBSIELLA OXYTOCA            1    90019.0   
21                                 KLEBSIELLA OXYTOCA            1    90022.0   
22                                 KLEBSIELLA OXYTOCA            1    90026.0   
23                                 KLEBSIELLA OXYTOCA            1    90028.0   
24                                 KLEBSIELLA OXYTOCA            1    90005.0   
76                                 KLEBSIELLA OXYTOCA            1    90008.0   
77                                 KLEBSIELLA OXYTOCA            1    90029.0   
90                 STAPHYLOCOCCUS, COAGULASE NEGATIVE            1        NaN   

                 AB_NAME DILUTION_TEXT DILUTION_COMPARISON  DILUTION_VALUE  \
13                                                                     NaN   
16            GENTAMICIN           <=1                  <=             1.0   
17            TOBRAMYCIN           <=1                  <=             1.0   
18           CEFTAZIDIME           <=1                  <=             1.0   
19           CEFTRIAXONE           <=1                  <=             1.0   
20         CIPROFLOXACIN        <=0.25                  <=             0.0   
21  AMPICILLIN/SULBACTAM          =>32                  =>            32.0   
22     PIPERACILLIN/TAZO         =>128                  =>           128.0   
23              CEFEPIME           <=1                  <=             1.0   
24             CEFAZOLIN          =>64                  =>            64.0   
76    TRIMETHOPRIM/SULFA           <=1                  <=             1.0   
77             MEROPENEM        <=0.25                  <=             0.0   
90                                                                     NaN   

   INTERPRETATION  
13                 
16          

In [19]:
print(f"There are {bloodcult.shape[0]} microbiology events on which a blood culture was performed")

There are 13 microbiology events on which a blood culture was performed


In [20]:
#display the number of urine specimen tested for bacterial growth.
urinecult = df_micro[df_micro['SPEC_TYPE_DESC'] == 'URINE']
urinecult

ROW_ID  SUBJECT_ID  HADM_ID            CHARTDATE            CHARTTIME  \
1   268038       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
2   268039       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
3   268040       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
4   268041       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
5   268042       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
6   268043       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
7   268044       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
8   268045       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
9   268046       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
10  268047       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
11  268048       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   
12  268049       40161   180396  2181-10-24 00:00:00  2181-10-24 18:26:00   

    SPEC_ITEMID SPEC_TYPE_DESC  ORG_ITEMID          ORG_NAME  ISOLATE_NUM  \
1         70079          URINE       80002  ESCHERICHIA COLI            1   
2         70079          URINE       80002  ESCHERICHIA COLI            1   
3         70079          URINE       80002  ESCHERICHIA COLI            1   
4         70079          URINE       80002  ESCHERICHIA COLI            1   
5         70079          URINE       80002  ESCHERICHIA COLI            1   
6         70079          URINE       80002  ESCHERICHIA COLI            1   
7         70079          URINE       80002  ESCHERICHIA COLI            1   
8         70079          URINE       80002  ESCHERICHIA COLI            1   
9         70079          URINE       80002  ESCHERICHIA COLI            1   
10        70079          URINE       80002  ESCHERICHIA COLI            1   
11        70079          URINE       80002  ESCHERICHIA COLI            1   
12        70079          URINE       80002  ESCHERICHIA COLI            1   

    AB_ITEMID               AB_NAME DILUTION_TEXT DILUTION_COMPARISON  \
1     90012.0            GENTAMICIN           <=1                  <=   
2     90029.0             MEROPENEM        <=0.25                  <=   
3     90028.0              CEFEPIME           <=1                  <=   
4     90022.0  AMPICILLIN/SULBACTAM             4                   =   
5     90019.0         CIPROFLOXACIN        <=0.25                  <=   
6     90018.0           CEFTRIAXONE           <=1                  <=   
7     90013.0            TOBRAMYCIN           <=1                  <=   
8     90010.0        NITROFURANTOIN            32                   =   
9     90008.0    TRIMETHOPRIM/SULFA           <=1                  <=   
10    90005.0             CEFAZOLIN           <=4                  <=   
11    90004.0            AMPICILLIN             8                   =   
12    90017.0           CEFTAZIDIME           <=1                  <=   

    DILUTION_VALUE INTERPRETATION  
1              1.0              S  
2              0.0              S  
3              1.0              S  
4              4.0              S  
5              0.0              S  
6              1.0              S  
7              1.0              S  
8             32.0              S  
9              1.0              S  
10             4.0              S  
11             8.0              S  
12             1.0              S

In [21]:
print(f"There are {urinecult.shape[0]} microbiology events on which a blood culture was performed")

There are 12 microbiology events on which a blood culture was performed


In [23]:
#display all the distinct drug values (from prescription table) used among all the records present.
drugs=pd.read_sql_query(""" SELECT DISTINCT DRUG FROM 'PRESCRIPTIONS'""",con)
drugs

DRUG
0                FoLIC Acid
1             Multivitamins
2                Carvedilol
3            Cyanocobalamin
4                Lisinopril
..                      ...
174           Mannitol 20 %
175  Cisatracurium Besylate
176              Gentamicin
177        Ciprofloxacin IV
178          NORepinephrine

[179 rows x 1 columns]

In [24]:
#considering the information available at transfers table, find out the numbers of patients admitted; transfered or discharged from the hospital.
# df_transfers = pd.read_sql_query("SELECT DISTINCT EVENTTYPE FROM 'TRANSFERS'",connection)
# df_transfers.head()
# #these shows us there are 3 distinct eventtypes 

print("10 patients have been admitted, as seen below")
print("20 patients have been transferred, as seen below")
print("10 patients have been discharged, as seen below")

#we can check this by seeing that there are 40 rows total:
rows = pd.read_sql_query("SELECT * FROM 'TRANSFERS'",con).shape[0]
print(f"total number of patients: {rows}")

10 patients have been admitted, as seen below
20 patients have been transferred, as seen below
10 patients have been discharged, as seen below
total number of patients: 40


In [26]:
#admissions:
pd.read_sql_query("SELECT COUNT(*) FROM 'TRANSFERS' WHERE EVENTTYPE = 'admit'",con)

COUNT(*)
0        10

In [27]:
#transfers:
pd.read_sql_query("SELECT COUNT(*) FROM 'TRANSFERS' WHERE EVENTTYPE = 'transfer'",con)

COUNT(*)
0        20

In [29]:
#discharges:
pd.read_sql_query("SELECT COUNT(*) FROM 'TRANSFERS' WHERE EVENTTYPE = 'discharge'",con)

COUNT(*)
0        10

In [30]:
#display only information for substances whose level of injection was higher than 100 mL (ordered based on the injection value) and caregiver id was 21452 and 21108. 

# TO BEEE DOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOONNNNNNNNNNNNNNNNEEEEEEEEEEEEEEE

In [32]:
#join admissions and patient table

admiss_patient = pd.read_sql_query("""SELECT * 
FROM 'ADMISSIONS'
INNER JOIN 'PATIENTS' ON ADMISSIONS.SUBJECT_ID=PATIENTS.SUBJECT_ID;""",con)
#We use an inner join, since I assume all patients will have a SUBJECT_ID

admiss_patient

#Here again, we can confirm that 
# admiss_patient.shape 
# has the same shape as 
# pd.read_sql_query("""SELECT * 
# FROM 'ADMISSIONS'""",connection).shape

ROW_ID SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME  \
0  39793      40036   198489  2141-08-01 23:46:00  2141-08-09 19:15:00   
1  39807      40080   162107  2106-05-31 16:43:00  2106-06-05 01:18:00   
2  39809      40084   195762  2173-01-31 22:11:00  2173-02-05 01:31:00   
3  39817      40116   157106  2150-02-19 00:12:00  2150-03-11 13:58:00   
4  39818      40120   146466  2120-01-27 20:41:00  2120-02-12 17:14:00   
5  39819      40120   158997  2120-02-21 15:22:00  2120-02-24 07:30:00   
6  39821      40124   126179  2130-02-04 02:26:00  2130-02-10 17:39:00   
7  39822      40124   146893  2130-08-12 05:49:00  2130-08-18 15:30:00   
8  39828      40161   180396  2181-10-24 15:58:00  2181-11-26 17:15:00   
9  39785      40004   121157  2171-03-01 17:51:00  2171-03-08 14:26:00   

             DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION  \
0                           EMERGENCY  CLINIC REFERRAL/PREMATURE   
1                           EMERGENCY  CLINIC REFERRAL/PREMATURE   
2  2173-02-05 01:31:00      EMERGENCY  CLINIC REFERRAL/PREMATURE   
3                           EMERGENCY  TRANSFER FROM HOSP/EXTRAM   
4                           EMERGENCY       EMERGENCY ROOM ADMIT   
5  2120-02-24 07:30:00      EMERGENCY       EMERGENCY ROOM ADMIT   
6                           EMERGENCY       EMERGENCY ROOM ADMIT   
7                           EMERGENCY       EMERGENCY ROOM ADMIT   
8  2181-11-26 17:15:00      EMERGENCY  TRANSFER FROM HOSP/EXTRAM   
9                           EMERGENCY  CLINIC REFERRAL/PREMATURE   

        DISCHARGE_LOCATION INSURANCE  ... HAS_IOEVENTS_DATA  \
0         HOME HEALTH CARE  Medicare  ...                 1   
1                      SNF  Medicaid  ...                 1   
2             DEAD/EXPIRED   Private  ...                 1   
3      SHORT TERM HOSPITAL  Medicaid  ...                 1   
4  LONG TERM CARE HOSPITAL  Medicare  ...                 1   
5             DEAD/EXPIRED  Medicare  ...                 1   
6                      SNF  Medicare  ...                 1   
7                      SNF  Medicare  ...                 1   
8             DEAD/EXPIRED  Medicare  ...                 1   
9                     HOME  Medicare  ...                 1   

  HAS_CHARTEVENTS_DATA ROW_ID SUBJECT_ID GENDER                  DOB  \
0                    1  30810      40036      M  2075-04-25 00:00:00   
1                    1  30821      40080      F  2027-08-04 00:00:00   
2                    1  30823      40084      M  2149-10-04 00:00:00   
3                    1  30828      40116      M  2109-03-21 00:00:00   
4                    1  30829      40120      M  2045-04-24 00:00:00   
5                    1  30829      40120      M  2045-04-24 00:00:00   
6                    1  30831      40124      F  2063-07-05 00:00:00   
7                    1  30831      40124      F  2063-07-05 00:00:00   
8                    1  30837      40161      M  2097-12-11 00:00:00   
9                    1  30802      40004      M  2118-09-11 00:00:00   

                   DOD             DOD_HOSP              DOD_SSN  EXPIRE_FLAG  
0  2143-09-03 00:00:00  2143-09-03 00:00:00  2143-09-03 00:00:00            1  
1  2106-06-14 00:00:00                       2106-06-14 00:00:00            1  
2  2173-02-05 00:00:00  2173-02-05 00:00:00  2173-02-05 00:00:00            1  
3  2150-09-26 00:00:00                       2150-09-26 00:00:00            1  
4  2120-02-24 00:00:00  2120-02-24 00:00:00                                 1  
5  2120-02-24 00:00:00  2120-02-24 00:00:00                                 1  
6  2130-11-03 00:00:00  2130-11-03 00:00:00  2130-11-03 00:00:00            1  
7  2130-11-03 00:00:00  2130-11-03 00:00:00  2130-11-03 00:00:00            1  
8  2181-11-26 00:00:00  2181-11-26 00:00:00  2181-11-26 00:00:00            1  
9  2172-11-12 00:00:00                       2172-11-12 00:00:00            1  

[10 rows x 28 columns]

In [13]:
#join admission, patient and icustays

In [14]:
#get the length of stay for all patients (measured in fractional days) and plot their distribution for male and female patients.

In [15]:
#select SUBJECT_ID, ICUSTAY_ID, from icustay and add a column 'LOS_nominal' that converts the LOS into a nominal variable ('Short' if los value > 3 else 'Long')

### Analysing patient's laboratory measurements

Table 'LABEVENTS' contains all laboratory measurements for some specific patients. 'D_LABITEMS' contains definitions for all ITEM_ID associated with lab measurements in the MIMIC database. In this part of the lab session, you should analyse the glucose levels for subject with id 40161. Query the database to retrive only the relevant information; plot the glucose levels for the given subject and interpret the results.

In [33]:
#query
query2 = """ SELECT  VALUENUM  FROM  LABEVENTS,D_LABITEMS
             WHERE SUBJECT_ID IS 40161 """ 

In [40]:
#select only the relevant subset of data
glucose_count = pd.read_sql_query(query2,con)
glucose_count.dropna() # Since we have NaN in our dataset so we are droping it 

VALUENUM
72         -11.0
73         -11.0
74         -11.0
75         -10.0
76         -10.0
...          ...
490555     590.0
490556     638.0
490557    1100.0
490558    5519.0
490559   20058.0

[476736 rows x 1 columns]

In [43]:
#plot
sns.barplot(data=glucose_count.dropna())

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/seaborn/matrix.py:649: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.
  warnings.warn(msg)


In [ ]:
#interpret result (written comments)

### Analysing patient's vital signs
During an intensive care unit (ICU) stay, the primary repository of a patient’s information is their electronic chart which displays patients’ routine vital signs,laboratory values, mental status, etc. 'CHARTEVENTS' table contains all the charted data available for a patient. 'ICUSTAYS' table (derived from the 'TRANSFERS' table) provides information about the date and time when the patient was transferred into/out the ICU. 'D_ITEMS' provides information about all items in the ICU databases. Considering the information you get from these three tables, find out how does the oxygen saturation level for patient with icu_stay_id 256307 change over time and interpret the results.

In [17]:
#query

#slect ony the relevant subset of data

#plot

#interpret result (written comments)